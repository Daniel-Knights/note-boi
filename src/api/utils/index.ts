export * from './decorators';
export * from './error';
export * from './http';
