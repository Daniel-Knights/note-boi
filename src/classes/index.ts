export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
