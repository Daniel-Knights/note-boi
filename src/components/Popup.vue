<template>
  <Teleport to="#app">
    <div id="popup">
      <div class="popup__content">
        <slot />
      </div>
    </div>
  </Teleport>
</template>

<script lang="ts" setup>
import { onBeforeUnmount } from 'vue';

const emit = defineEmits(['close']);

function closePopup() {
  document.removeEventListener('keydown', keyboardCloseHandler);
  document.removeEventListener('mousedown', clickCloseHandler);

  emit('close');
}

function keyboardCloseHandler(event: KeyboardEvent) {
  if (event.key === 'Escape') closePopup();
}
function clickCloseHandler(event: MouseEvent) {
  if (!(event.target as HTMLElement)?.closest('.popup__content')) {
    closePopup();
  }
}

// setTimeout prevents immediate close
setTimeout(() => {
  document.addEventListener('keydown', keyboardCloseHandler);
  document.body.addEventListener('mousedown', clickCloseHandler);
});

onBeforeUnmount(closePopup);
</script>

<style lang="scss" scoped>
#popup {
  @include v.flex-x(center, center);
  position: fixed;
  @include v.cover;
  @include v.equal-dimensions(100%);
  background-color: rgba(0, 0, 0, 0.5);
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  z-index: 100;
}

.popup__content {
  padding: 1em;
  max-width: 400px;
  color: var(--colour__primary);
  background-color: var(--colour__secondary);
}
</style>
