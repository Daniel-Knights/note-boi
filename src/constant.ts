export const NOTE_EVENTS = {
  new: 'note-new',
  /** For preventing note edit on text change. */
  select: 'note-select',
  /** For whenever the content of a note is changed. */
  change: 'note-change',
  unsynced: 'note-unsynced',
};

export const STORAGE_KEYS = {
  USERNAME: 'username',
  TOKEN: 'token',
  UNSYNCED: 'unsynced-note-ids',
  THEME: 'theme',
  MENU_WIDTH: 'note-menu-width',
};
