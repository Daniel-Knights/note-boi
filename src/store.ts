import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
