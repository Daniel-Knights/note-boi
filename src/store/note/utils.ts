import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
