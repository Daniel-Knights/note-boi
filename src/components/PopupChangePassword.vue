<template>
  <Popup @close="emit('close')">
    <div id="change-password" data-test-id="popup-change-password">
      <h2 data-test-id="heading">Change Password</h2>
      <form @submit.prevent="handleSubmit" class="form" data-test-id="form">
        <input
          v-model="syncState.password"
          @input="validation.currentPassword = true"
          class="form__input"
          :class="{ 'form__input--invalid': !validation.currentPassword }"
          type="password"
          placeholder="Current Password"
          ref="currentPassword"
          data-test-id="current-password"
        />
        <input
          v-model="syncState.newPassword"
          @input="validation.newPassword = true"
          class="form__input"
          :class="{ 'form__input--invalid': !validation.newPassword }"
          type="password"
          placeholder="New Password"
          data-test-id="new-password"
        />
        <input
          v-model="confirmNewPassword"
          @input="validation.confirmNewPassword = true"
          class="form__input"
          :class="{ 'form__input--invalid': !validation.confirmNewPassword }"
          type="password"
          placeholder="Confirm New Password"
          data-test-id="confirm-new-password"
        />
        <p
          v-if="syncState.appError.display?.form"
          class="form__error"
          data-test-id="error-message"
        >
          {{ syncState.appError.message || 'Something went wrong' }}
        </p>
        <input type="submit" value="Submit" class="button button--default" />
      </form>
    </div>
  </Popup>
</template>

<script lang="ts" setup>
import { onMounted, reactive, ref } from 'vue';

import { AppError, ERROR_CODE } from '../classes';
import { MIN_PASSWORD_LENGTH } from '../constant';
import { changePassword, syncState } from '../store/sync';

import Popup from './Popup.vue';

const emit = defineEmits(['close']);

const currentPassword = ref<HTMLInputElement>();
const confirmNewPassword = ref('');

const validation = reactive({
  currentPassword: true,
  newPassword: true,
  confirmNewPassword: true,
});

async function handleSubmit() {
  validation.currentPassword = !!syncState.password;
  validation.newPassword = !!syncState.newPassword;
  validation.confirmNewPassword = !!confirmNewPassword.value;

  if (Object.values(validation).some((v) => v === false)) {
    return;
  }

  if (syncState.newPassword.length < MIN_PASSWORD_LENGTH) {
    syncState.appError = new AppError({
      code: ERROR_CODE.FORM_VALIDATION,
      message: `Password must be at least ${MIN_PASSWORD_LENGTH} characters long`,
      display: {
        form: true,
      },
    });

    return;
  }

  if (confirmNewPassword.value !== syncState.newPassword) {
    syncState.appError = new AppError({
      code: ERROR_CODE.FORM_VALIDATION,
      message: "Passwords don't match",
      display: {
        form: true,
      },
    });

    return;
  }

  if (syncState.newPassword === syncState.password) {
    syncState.appError = new AppError({
      code: ERROR_CODE.FORM_VALIDATION,
      message: 'Current and new passwords must be different',
      display: {
        form: true,
      },
    });

    return;
  }

  await changePassword();

  if (syncState.appError.isNone) {
    confirmNewPassword.value = '';

    emit('close');
  }
}

onMounted(() => {
  currentPassword.value?.focus();
});
</script>

<style lang="scss" scoped></style>
