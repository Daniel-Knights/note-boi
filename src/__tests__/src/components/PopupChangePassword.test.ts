import { mount } from '@vue/test-utils';

import * as a from '../../../api';
import * as s from '../../../store/sync';
import { AppError, ERROR_CODE } from '../../../classes';
import { MIN_PASSWORD_LENGTH } from '../../../constant';
import { clearMockApiResults, mockApi } from '../../mock';
import { assertAppError, assertRequest, getByTestId, waitUntil } from '../../utils';

import Popup from '../../../components/Popup.vue';
import PopupChangePassword from '../../../components/PopupChangePassword.vue';

function mountPopupChangePassword() {
  return mount(PopupChangePassword, {
    global: {
      stubs: { teleport: true },
    },
  });
}

describe('PopupChangePassword', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mountPopupChangePassword();

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);
  });

  it('Emits close', () => {
    const wrapper = mountPopupChangePassword();
    wrapper.getComponent(Popup).vm.$emit('close');

    assert.lengthOf(wrapper.emitted('close')!, 1);
  });

  it('Shows error message', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mountPopupChangePassword();

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
      message: 'Mock error',
      display: { form: true },
    });

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);

    const errorMessageWrapper = getByTestId(wrapper, 'error-message');
    assert.strictEqual(errorMessageWrapper.text(), 'Mock error');
  });

  it('Shows default error message when none provided', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mountPopupChangePassword();

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
      display: { form: true },
    });

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);

    const errorMessageWrapper = getByTestId(wrapper, 'error-message');
    assert.strictEqual(errorMessageWrapper.text(), 'Something went wrong');
  });

  it('Validates fields and submits', async () => {
    const { calls } = mockApi();

    s.syncState.username = 'd';
    s.syncState.password = '1';

    await a.login();

    const wrapper = mountPopupChangePassword();
    const wrapperVm = wrapper.vm as unknown as {
      confirmNewPassword: string;
      validation: {
        currentPassword: boolean;
        newPassword: boolean;
        confirmNewPassword: boolean;
      };
    };

    assert.isTrue(wrapper.isVisible());

    const currentPasswordInput = getByTestId<HTMLInputElement>(
      wrapper,
      'current-password'
    );
    const newPasswordInput = getByTestId<HTMLInputElement>(wrapper, 'new-password');
    const confirmNewPasswordInput = getByTestId<HTMLInputElement>(
      wrapper,
      'confirm-new-password'
    );

    assert.isEmpty(currentPasswordInput.element.value);
    assert.isEmpty(newPasswordInput.element.value);
    assert.isEmpty(confirmNewPasswordInput.element.value);
    assert.isEmpty(s.syncState.password);
    assert.isEmpty(s.syncState.newPassword);
    assert.isEmpty(wrapperVm.confirmNewPassword);
    assert.isTrue(wrapperVm.validation.currentPassword);
    assert.isTrue(wrapperVm.validation.newPassword);
    assert.isTrue(wrapperVm.validation.confirmNewPassword);

    const changePasswordSpy = vi.spyOn(a, 'changePassword');

    const formWrapper = getByTestId(wrapper, 'form');
    await formWrapper.trigger('submit');

    expect(changePasswordSpy).not.toHaveBeenCalled();

    assertAppError();
    assert.isFalse(wrapperVm.validation.currentPassword);
    assert.isFalse(wrapperVm.validation.newPassword);
    assert.isFalse(wrapperVm.validation.confirmNewPassword);

    await currentPasswordInput.setValue('1');

    assertAppError();
    assert.isTrue(wrapperVm.validation.currentPassword);
    assert.isFalse(wrapperVm.validation.newPassword);
    assert.isFalse(wrapperVm.validation.confirmNewPassword);

    newPasswordInput.setValue('2');

    assertAppError();
    assert.isTrue(wrapperVm.validation.currentPassword);
    assert.isTrue(wrapperVm.validation.newPassword);
    assert.isFalse(wrapperVm.validation.confirmNewPassword);

    confirmNewPasswordInput.setValue('2');

    assertAppError();
    assert.isTrue(wrapperVm.validation.currentPassword);
    assert.isTrue(wrapperVm.validation.newPassword);
    assert.isTrue(wrapperVm.validation.confirmNewPassword);

    await formWrapper.trigger('submit');

    assertAppError({
      code: ERROR_CODE.FORM_VALIDATION,
      message: `Password must be at least ${MIN_PASSWORD_LENGTH} characters long`,
      display: { form: true },
    });

    newPasswordInput.setValue('123456');

    await formWrapper.trigger('submit');

    assertAppError({
      code: ERROR_CODE.FORM_VALIDATION,
      message: "Passwords don't match",
      display: { form: true },
    });

    confirmNewPasswordInput.setValue('123456');

    expect(changePasswordSpy).not.toHaveBeenCalled();

    assert.isTrue(wrapperVm.validation.currentPassword);
    assert.isTrue(wrapperVm.validation.newPassword);
    assert.isTrue(wrapperVm.validation.confirmNewPassword);
    assert.isNotEmpty(s.syncState.username);
    assert.isTrue(s.syncState.isLoggedIn);
    assert.isNotEmpty(s.syncState.password);
    assert.isNotEmpty(s.syncState.newPassword);

    clearMockApiResults({ calls });

    await formWrapper.trigger('submit');
    await waitUntil(() => !s.syncState.loadingCount);

    expect(changePasswordSpy).toHaveBeenCalledOnce();

    assertAppError();
    assert.isEmpty(wrapperVm.confirmNewPassword);
    assert.lengthOf(wrapper.emitted('close')!, 1);
    assert.strictEqual(calls.size, 3);
    assert.isTrue(calls.request.has('/account/change-password'));
    assertRequest('/account/change-password', calls.request[0]!.calledWith!);
    assert.isTrue(calls.invoke.has('get_access_token'));
    assert.deepEqual(calls.invoke[0]!.calledWith, { username: 'd' });
    assert.isTrue(calls.invoke.has('set_access_token'));
    assert.deepEqual(calls.invoke[1]!.calledWith, {
      username: 'd',
      accessToken: 'test-token',
    });
  });
});
