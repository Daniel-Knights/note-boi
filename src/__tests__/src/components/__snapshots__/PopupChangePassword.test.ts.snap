// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopupChangePassword > Emits close > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupChangePassword > Emits close > allCalls 1`] = `[]`;

exports[`PopupChangePassword > Emits close > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupChangePassword > Emits close > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupChangePassword > Emits close > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupChangePassword > Emits close > openedPopup 1`] = `undefined`;

exports[`PopupChangePassword > Emits close > selectedTheme 1`] = `"System"`;

exports[`PopupChangePassword > Emits close > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupChangePassword > Emits close > storage.get('THEME') 1`] = `null`;

exports[`PopupChangePassword > Emits close > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupChangePassword > Emits close > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupChangePassword > Emits close > storage.get('USERNAME') 1`] = `null`;

exports[`PopupChangePassword > Emits close > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupChangePassword > Emits close > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupChangePassword > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupChangePassword > Mounts > allCalls 1`] = `[]`;

exports[`PopupChangePassword > Mounts > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupChangePassword > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupChangePassword > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupChangePassword > Mounts > openedPopup 1`] = `undefined`;

exports[`PopupChangePassword > Mounts > selectedTheme 1`] = `"System"`;

exports[`PopupChangePassword > Mounts > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupChangePassword > Mounts > storage.get('THEME') 1`] = `null`;

exports[`PopupChangePassword > Mounts > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupChangePassword > Mounts > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupChangePassword > Mounts > storage.get('USERNAME') 1`] = `null`;

exports[`PopupChangePassword > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupChangePassword > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupChangePassword > Shows default error message when none provided > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupChangePassword > Shows default error message when none provided > allCalls 1`] = `[]`;

exports[`PopupChangePassword > Shows default error message when none provided > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupChangePassword > Shows default error message when none provided > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupChangePassword > Shows default error message when none provided > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupChangePassword > Shows default error message when none provided > openedPopup 1`] = `undefined`;

exports[`PopupChangePassword > Shows default error message when none provided > selectedTheme 1`] = `"System"`;

exports[`PopupChangePassword > Shows default error message when none provided > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupChangePassword > Shows default error message when none provided > storage.get('THEME') 1`] = `null`;

exports[`PopupChangePassword > Shows default error message when none provided > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupChangePassword > Shows default error message when none provided > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupChangePassword > Shows default error message when none provided > storage.get('USERNAME') 1`] = `null`;

exports[`PopupChangePassword > Shows default error message when none provided > syncState 1`] = `
{
  "appError": AppError {
    "code": 5,
    "display": {
      "form": true,
    },
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupChangePassword > Shows default error message when none provided > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupChangePassword > Shows error message > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupChangePassword > Shows error message > allCalls 1`] = `[]`;

exports[`PopupChangePassword > Shows error message > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupChangePassword > Shows error message > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupChangePassword > Shows error message > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupChangePassword > Shows error message > openedPopup 1`] = `undefined`;

exports[`PopupChangePassword > Shows error message > selectedTheme 1`] = `"System"`;

exports[`PopupChangePassword > Shows error message > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupChangePassword > Shows error message > storage.get('THEME') 1`] = `null`;

exports[`PopupChangePassword > Shows error message > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupChangePassword > Shows error message > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupChangePassword > Shows error message > storage.get('USERNAME') 1`] = `null`;

exports[`PopupChangePassword > Shows error message > syncState 1`] = `
{
  "appError": AppError {
    "code": 5,
    "display": {
      "form": true,
    },
    "message": "Mock error",
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupChangePassword > Shows error message > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupChangePassword > Validates fields and submits > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`PopupChangePassword > Validates fields and submits > allCalls 1`] = `
[
  [
    "request",
    {
      "name": "/login",
    },
  ],
  [
    "emits",
    {
      "calledWith": {
        "data": {
          "is_logged_in": true,
        },
        "isFrontendEmit": true,
      },
      "name": "auth",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "note": {
          "content": {
            "body": "",
            "delta": {
              "ops": [],
            },
            "title": "",
          },
          "id": "id",
          "timestamp": 0,
        },
      },
      "name": "new_note",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "notes": [
          {
            "content": {
              "body": "",
              "delta": {
                "ops": [],
              },
              "title": "",
            },
            "id": "id",
            "timestamp": 0,
          },
        ],
      },
      "name": "sync_local_notes",
    },
  ],
  [
    "request",
    {
      "name": "/account/password/change",
    },
  ],
]
`;

exports[`PopupChangePassword > Validates fields and submits > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupChangePassword > Validates fields and submits > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupChangePassword > Validates fields and submits > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [
    {
      "content": {
        "body": "",
        "delta": {
          "ops": [],
        },
        "title": "",
      },
      "id": "id",
      "timestamp": 0,
    },
  ],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupChangePassword > Validates fields and submits > openedPopup 1`] = `undefined`;

exports[`PopupChangePassword > Validates fields and submits > selectedTheme 1`] = `"System"`;

exports[`PopupChangePassword > Validates fields and submits > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupChangePassword > Validates fields and submits > storage.get('THEME') 1`] = `null`;

exports[`PopupChangePassword > Validates fields and submits > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupChangePassword > Validates fields and submits > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupChangePassword > Validates fields and submits > storage.get('USERNAME') 1`] = `"d"`;

exports[`PopupChangePassword > Validates fields and submits > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": true,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "d",
}
`;

exports[`PopupChangePassword > Validates fields and submits > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
